internal/provider/audit_resource.go

```diff
@@ -0,0 +1,227 @@
+		Attributes: map[string]schema.Attribute{
+			"role": schema.StringAttribute{
+				Description: "Name of the role.",
+				Required:    true,
+			},
+			"audit_log_option": schema.StringAttribute{
+				Description: "Value for the pgaudit.log option for this role. Examples: 'none', 'all', 'ddl', 'write', etc.",
+				Required:    true,
+			},
+		},
+	}
+}
+
+type auditModel struct {
+	Role           string `tfsdk:"role"`
+	AuditLogOption string `tfsdk:"audit_log_option"`
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *auditResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Add a nil check when handling ProviderData because Terraform
+	// sets that data after it calls the ConfigureProvider RPC.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(F)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected provider.F, got %T", req.ProviderData),
+		)
+	}
+
+	r.getDB = client
+}
+
+// Create creates the resource and sets the initial Terraform state.
+func (r *auditResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	// Retrieve value from plan
+	var plan auditModel
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create the resource
+	sqlstr := sqlSetAuditLog(plan.Role, plan.AuditLogOption)
+
+	db, err := r.getDB(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Failed to get database connection",
+			"Failed to get database connection: "+err.Error(),
+		)
+		return
+	}
+	defer db.Close()
+	if _, err = db.ExecContext(ctx, sqlstr); err != nil {
+		resp.Diagnostics.AddError(
+			"Failed to execute SQL",
+			"Failed to execute SQL: "+err.Error(),
+		)
+		return
+	}
+
+	// Set state to fully populated data
+	diags = resp.State.Set(ctx, plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Read refreshes the Terraform state with the latest data.
+func (r *auditResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	// Get the current state
+	var state auditModel
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get the actual value in postgres
+	db, err := r.getDB(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Failed to get database connection",
+			"Failed to get database connection: "+err.Error(),
+		)
+		return
+	}
+	defer db.Close()
+
+	var auditLogOption string
+	if err := db.QueryRowContext(ctx, "SELECT pg_catalog.current_setting('pgaudit.log') FROM pg_roles WHERE rolname = $1;", state.Role).Scan(&auditLogOption); err != nil {
```

Comment:
`pg_catalog.current_setting('pgaudit.log')` returns the value for the *user running the query*, not the user we want to search :rofl:

Reply:
i've fixed this in my commit